test/integration/replicaset/replicaset_test.go

```diff
@@ -515,3 +519,682 @@ func TestRSSelectorImmutability(t *testing.T) {
+	if err != nil {
+		t.Fatalf("failed to parse ReplicaSet selector: %v", err)
+	}
+	var readyPods int32
+	err = wait.Poll(1*time.Second, 1*time.Minute, func() (bool, error) {
+		readyPods = 0
+		pods, err := clientSet.Core().Pods(ns).List(metav1.ListOptions{LabelSelector: selector.String()})
+		if err != nil {
+			t.Logf("failed to list ReplicaSet pods, will retry later: %v", err)
+			return false, nil
+		}
+		for i := range pods.Items {
+			pod := pods.Items[i]
+			if podutil.IsPodReady(&pod) {
+				readyPods++
+				continue
+			}
+			addPodConditionReady(&pod, metav1.Now())
+			if _, err = clientSet.Core().Pods(ns).UpdateStatus(&pod); err != nil {
+				t.Logf("failed to update ReplicaSet pod %s, will retry later: %v", pod.Name, err)
+			} else {
+				readyPods++
+			}
+		}
+		if readyPods >= *rs.Spec.Replicas {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to mark all ReplicaSet pods to ready: %v", err)
+	}
+	//fmt.Println("readyPods: %d, rs.spec.replicas: %d", readyPods, *rs.Spec.Replicas)
+}
+
+func waitAllPodsReady(t *testing.T, clientSet clientset.Interface, rs *v1beta1.ReplicaSet, ns string) {
+	rsClient := clientSet.Extensions().ReplicaSets(ns)
+	if err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		updatedRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if updatedRS.Status.ReadyReplicas != *rs.Spec.Replicas {
+			return false, nil
+		} else {
+			return true, nil
+		}
+	}); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestReadyAndAvailableReplicas(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-ready-and-available-replicas", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 3)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	// Make .spec.minReadySeconds long enough so that we can test the replicas are unavailable initially
+	rs.Spec.MinReadySeconds = 3600
+
+	// Create 3 pods
+	var podList []*v1.Pod
+	moment := metav1.Now()
+	for i := 0; i < 3; i++ {
+		// When creating the pod, manually set pod status as there is no kubelet in integration test env
+		pod := newPod(fmt.Sprintf("pod-%d", i+1), rs, v1.PodRunning, &moment, true)
+		pod.Labels = labelMap
+		pod.Spec.Containers = []v1.Container{
+			{
+				Name:  "nginx",
+				Image: "gcr.io/google_containers/nginx-slim:0.8",
+			},
+		}
+		podList = append(podList, pod)
+	}
+	createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, podList, ns.Name)
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	markAllPodsReady(t, c, rs, ns.Name)
+	waitAllPodsReady(t, c, rs, ns.Name)
+	waitRSStable(t, c, rs, ns.Name)
+
+	rs, err := c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(rs.Name, metav1.GetOptions{})
```

Comment:
You already checked these things via `waitAllPodsReady` and `waitRSStable` above.

Reply:
Yes, this check is unnecessary.